netket/nn/linear.py

```diff
@@ -347,6 +347,115 @@ def create_DenseSymm(
     return DenseSymm(symmetries=perm_fn, *args, **kwargs)
 
 
+class DenseEquivariant(Module):
+    """An equivariant layer that takes a feature map with embeddings corresponding to various
+    poses of a symmetry group, shape = [n_symm*in_features] and outputs a feature map of identical
+    shape [n_symm*out_features]. This transformation respects the group structure such that
+    g L({f}) = L(g{f}) where g is an element of the symmetry group and {f} are the feature embeddings.
+    """
+
+    group_algebra: Tuple
+    """List of tuples generated by graph.group_algebra(automorphisms) that specifies the mapping
+    between absolute and relative poses """
+    in_features: int
+    """The number of symmetry-reduced input features. The full input size
+    is n_symm*in_features."""
+    out_features: int
+    """The number of symmetry-reduced output features. The full output size
+    is n_symm*out_features."""
+    use_bias: bool = True
+    """Whether to add a bias to the output (default: True)."""
+    dtype: Any = jnp.float64
+    """The dtype of the weights."""
+    precision: Any = None
+    """numerical precision of the computation see `jax.lax.Precision`for details."""
+
+    kernel_init: Callable[[PRNGKey, Shape, Dtype], Array] = default_kernel_init
+    """Initializer for the Dense layer matrix."""
+    bias_init: Callable[[PRNGKey, Shape, Dtype], Array] = zeros
+    """Initializer for the bias."""
+
+    def setup(self):
+        self.n_symm = int(np.sqrt(len(self.group_algebra)))
+
+    def full_kernel(self, kernel):
+        """
+        Converts the symmetry-reduced kernel of shape (n_sites, features) to
+        the full Dense kernel of shape (n_sites, features * n_symm).
+        """
+
+        result = jnp.take(kernel, self.group_algebra, 0)
+        result = result.reshape(
+            self.n_symm, self.n_symm, self.in_features, self.out_features
+        )
+        result = result.transpose(2, 0, 3, 1).reshape(
+            self.n_symm * self.in_features, -1
+        )
```

Comment:
I was under the impression that reshapes copy memory. At least that is the case for Pytorch where .view() was the analog of reshape that doesn't touch memory. I trust your read on the situation.

I can't think of a good way to avoid the transposition. The mapping is transformed_kernel[a*n_symm + i, b*n_symm + j] = kernel[group_algebra[i,j], a, b]

Reply:
right now group_algebra is a flattened and turned into a tuple

Reply:
I can't speak for torch as I never used it, but that is the case* in jax. (*: Technically it's not guaranteed as it's treated like an optimization, but the Jax compiler is very good at optimizing away memory reuses)

Reply:
I tested tensordot with and without transpose, as well as jax.lax.dot and they all take about 3 microseconds per iteration on a Colab GPU. The intra run variance for N=1000 was larger than the individual differences between runs  

```
n_sym = 100
n_hid = 16
batch_size = 100

def transpose(inp,kernel):
    kernel = kernel.transpose(2,0,3,1)
    kernel = kernel.reshape(n_hid*n_sym,n_hid*n_sym)

    out = jax.lax.dot_general(inp,kernel,(((inp.ndim - 1,), (0,)), ((), ())))

    return out

def lax_dot(inp,kernel):
    kernel = kernel.transpose(2,0,3,1)
    kernel = kernel.reshape(n_hid*n_sym,n_hid*n_sym)

    out = jax.lax.dot(inp,kernel)

    return out

def no_transpose(inp,kernel):
    inp = inp.reshape(batch_size,n_hid,n_sym) 

    out = jax.lax.dot_general(inp,kernel,(((1,2), (2,0)), ((), ())))
    out = out.reshape(batch_size,n_sym*n_hid)

    return out

jit_transpose = jax.jit(partial(transpose,kernel=kernel_unshaped))
jit_no_transpose = jax.jit(partial(no_transpose,kernel=kernel_unshaped))
jit_dot = jax.jit(partial(lax_dot,kernel=kernel_unshaped))
```

Reply:
I think the XLA compiler is compiling the same code for both. So that is good.